### PR TITLE
[wip] a regexp used to filter namespaces

### PR DESCRIPTION
**Pull request by honnix**
# TL;DR
When watching "all" namespaces, a regexp filter can be used to narrow down the scope.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added (partially)
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
**This PR now depends on https://github.com/lyft/flytestdlib/pull/58.**

It is already possible to watch one specific namespace, however it seems not possible to watch
multiple ones according to https://github.com/kubernetes/client-go/issues/580

This PR added a new config (regexp) that will be used to ensure only workflow from interested namespace(s) is processed: informer callback and garbage collector.

This is probably the first time I wrote this much `go` code, please help me improve. Also I haven't spent much time working on `controller` test cases (which seems to be missing heavily).

## Tracking Issue
NA (will create one better context/use case is needed)

## Follow-up issue
NA

### CODE REVIEW

File: pkg/controller/config/config.go [hunk 1]
```diff
@@ -59,6 +59,7 @@ type Config struct {
 	WorkflowReEval      config.Duration      `json:"workflow-reeval-duration" pflag:"\"30s\",Frequency of re-evaluating workflows"`
 	DownstreamEval      config.Duration      `json:"downstream-eval-duration" pflag:"\"60s\",Frequency of re-evaluating downstream tasks"`
 	LimitNamespace      string               `json:"limit-namespace" pflag:"\"all\",Namespaces to watch for this propeller"`
+	NamespaceFilter     string               `json:"namespace-filter" pflag:",Regular expression to apply fine grain filter when watching all namespaces"`
```
Comment (EngHabu) [inline pkg/controller/config/config.go]: I would ideally create a type like [this](https://github.com/lyft/flytestdlib/blob/master/config/url.go) to be able to get config validation (using `propeller config validate`) before actually running propeller...

Comment (EngHabu) [inline pkg/controller/config/config.go]:
```suggestion
	NamespaceFilterRegex     string               `json:"namespace-filter" pflag:",Regular expression to apply fine grain filter when watching all namespaces"`
```

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -53,9 +54,10 @@ type Controller struct {
 	workflowStore       workflowstore.FlyteWorkflow
 	// recorder is an event recorder for recording Event resources to the
 	// Kubernetes API.
-	recorder      record.EventRecorder
-	metrics       *metrics
-	leaderElector *leaderelection.LeaderElector
+	recorder         record.EventRecorder
+	metrics          *metrics
+	leaderElector    *leaderelection.LeaderElector
+	namespaceFilterR *regexp.Regexp
```
Comment (EngHabu) [inline pkg/controller/controller.go]:
```suggestion
	namespaceFilterRegex *regexp.Regexp
```

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -220,12 +238,24 @@ func New(ctx context.Context, cfg *config.Config, kubeclientset kubernetes.Inter
 		wfLauncher = launchplan.NewFailFastLaunchPlanExecutor()
 	}
 
+	namespaceFilterR := func() *regexp.Regexp {
+		if cfg.NamespaceFilter != "" {
+			r, err := regexp.Compile(cfg.NamespaceFilter)
+			if err != nil {
+				logger.Warningf(ctx, "[%s] is not a valid regular expression, skipping namespace filtering. Error : [%v]", cfg.NamespaceFilter, err)
```
Comment (EngHabu) [inline pkg/controller/controller.go]: This should return an error and not try to recover from bad config errors.

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -220,12 +238,24 @@ func New(ctx context.Context, cfg *config.Config, kubeclientset kubernetes.Inter
 		wfLauncher = launchplan.NewFailFastLaunchPlanExecutor()
 	}
 
+	namespaceFilterR := func() *regexp.Regexp {
```
Comment (EngHabu) [inline pkg/controller/controller.go]: Is there a reason for this predicate?

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -220,12 +238,24 @@ func New(ctx context.Context, cfg *config.Config, kubeclientset kubernetes.Inter
...
+		if cfg.NamespaceFilter != "" {
+			r, err := regexp.Compile(cfg.NamespaceFilter)
```
Comment (EngHabu) [inline pkg/controller/controller.go]: If you opt for the config type (first comment), you are guaranteed a valid regex at this point...

File: pkg/controller/controller.go (hunk 4 above)
Comment (honnix) [reply pkg/controller/controller.go]: I had hard time dealing with `nil`. What I want was:

```
a := nil

if b {
    a = "something not nil"
}
```

It's my bad `go` skill I believe.

**Comment by honnix** [comment]
@EngHabu Yeah, client side filtering is definitely not ideal. It's good to know that you have been experimenting with a replacement. I remember @kumare3 mentioned that when we talked last time. It's been a while and I probably forgot about the details. 😅 

Our immediate use case is actually not for sharding, partitioning or in general for performance sake.

Imagine we have a production setup with all the components running and propeller watching `all`; we would like to start a propeller locally that can utilise everything in the production (admin, console, k8s cluster, etc.) but only watches a debug namespace. With the production propeller watching `all`, we cannot route the workflow execution to the local propeller only.

File: pkg/controller/controller.go (hunk 3 above)
Comment (honnix) [reply pkg/controller/controller.go]: Yeah that might be better. I wasn't so sure whether this should be treated as failure. This could be done better as suggested by https://github.com/lyft/flytepropeller/pull/90#discussion_r393113304

File: pkg/controller/config/config.go (hunk 1 above)
Comment (honnix) [reply pkg/controller/config/config.go]: Will fix in a commit for code alignment sake.

File: pkg/controller/controller.go (hunk 2 above)
Comment (honnix) [reply pkg/controller/controller.go]: Will fix in a commit for code alignment sake.

**Comment by honnix** [comment]
That being said, please do not consider this as must to have feature. The PR was more my exercise while reading through propeller code base.

File: pkg/controller/config/config.go (hunk 1 above)
Comment (honnix) [reply pkg/controller/config/config.go]: Thanks for the suggestion. I tried it in https://github.com/lyft/flytestdlib/pull/58.

**Comment by honnix** [comment]
It would be a lot better this filtering is done on server side than client. Closing this PR for now.
